clients/drcachesim/tests/window_test.cpp

```diff
@@ -0,0 +1,185 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test for multi-window traces where it inserts sleep between windows and
+ * checks that the timestamp for second buffer in a new window is closer to the
+ * previous timestamp than the sleep time.
+ */
+
+#include <assert.h>
+#include <chrono>
+#include <iostream>
+#include <string.h>
+#include <thread>
+#include "analysis_tool.h"
+#include "dr_api.h"
+#include "scheduler.h"
+#include "tracer/raw2trace.h"
+#include "tracer/raw2trace_directory.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+namespace {
+
+constexpr int SECONDS_TO_SLEEP = 4;
+constexpr int SECONDS_TO_TIMESTAMP = 1000000;
+
+} // namespace
+
+bool
+my_setenv(const char *var, const char *value)
+{
+#ifdef UNIX
+    return setenv(var, value, 1 /*override*/) == 0;
+#else
+    return SetEnvironmentVariable(var, value) == TRUE;
+#endif
+}
+
+int
+fib_with_sleep(int n)
+{
+    dr_fprintf(STDERR, "Calculating fibonacci of: %d\n", n);
+    std::this_thread::sleep_for(std::chrono::seconds(SECONDS_TO_SLEEP));
+    if (n <= 1)
+        return 1;
+    return fib_with_sleep(n - 1) + fib_with_sleep(n - 2);
+}
+
+static std::string
+post_process()
+{
+    const char *raw_dir;
+    drmemtrace_status_t mem_res = drmemtrace_get_output_path(&raw_dir);
+    assert(mem_res == DRMEMTRACE_SUCCESS);
+    std::string outdir = std::string(raw_dir) + DIRSEP + "trace";
+    void *dr_context = dr_standalone_init();
+    {
+        raw2trace_directory_t dir;
+        if (!dr_create_dir(outdir.c_str())) {
+            std::cerr << "Failed to create output dir";
+            assert(false);
+        }
+        std::string dir_err = dir.initialize(raw_dir, outdir);
+        assert(dir_err.empty());
+        raw2trace_t raw2trace(dir.modfile_bytes_, dir.in_files_, dir.out_files_,
+                              dir.out_archives_, dir.encoding_file_,
+                              dir.serial_schedule_file_, dir.cpu_schedule_file_,
+                              dr_context,
+                              0
+#ifdef WINDOWS
+                              /* TODO i#3983: Creating threads in standalone mode
+                               * causes problems.  We disable the pool for now.
+                               */
+                              ,
+                              0
+#endif
+        );
+        std::string error = raw2trace.do_conversion();
+        if (!error.empty()) {
+            std::cerr << "raw2trace failed: " << error << "\n";
+            assert(false);
+        }
+    }
+    dr_standalone_exit();
+    return outdir;
+}
+
+static std::string
+gather_trace()
+{
+    std::string dr_ops(
+        "-stderr_mask 0xc -client_lib ';;-offline -offline -trace_after_instrs 1000 "
+        "-trace_for_instrs 2500 -retrace_every_instrs 1000");
```

Comment:
Comment on how these values were picked: related to the sleep time?

Reply:
Done.